File: src/main/java/org/kantega/niagara/task/TaskExecutor.java
package org.kantega.niagara.task;

import fj.Unit;
import org.kantega.niagara.Try;

import java.time.Duration;
import java.time.Instant;
import java.util.concurrent.*;
import java.util.function.Consumer;
import java.util.function.Function;

public class TaskExecutor {

    final ForkJoinPool executorService =
      ForkJoinPool.commonPool();

    final ScheduledExecutorService scheduledExecutorService =
      Executors.newSingleThreadScheduledExecutor();

    final Function<Throwable, Task<Unit>> defaultHandler =
      t -> Task.run(t::printStackTrace);


    public <A> void eval(Task<A> task) {
        enqueue(
          new TaskContext(this),
          task,
          aTry -> aTry.doEffect(t -> eval(defaultHandler.apply(t)), a -> {}));
    }

    public <A> void enqueue(TaskContext tc, Task<A> t, Consumer<Try<A>> continuation) {
        executorService.submit(() -> {
            try {
                t.perform(tc, continuation);
            } catch (Throwable e) {
                continuation.accept(Try.fail(e));
            }
        });
    }

    public <A> CompletableFuture<Unit> enqueueStage(TaskContext tc, Task<A> t, Consumer<Try<A>> continuation) {
        var cf = new CompletableFuture<Unit>();
        executorService.submit(() -> {
            try {
                t.perform(tc, continuation);
                cf.complete(Unit.unit());
            } catch (Throwable e) {
                continuation.accept(Try.fail(e));
            }
        });
        return cf;
    }

    public <A> void schedule(TaskContext tc, Task<A> t, Consumer<Try<A>> continuation, Instant at) {
        Duration d = Duration.between(Instant.now(), at);
        if (d.isNegative())
            enqueue(tc, t, continuation);
        else
            scheduledExecutorService.schedule(() -> enqueue(tc, t, continuation), d.toMillis(), TimeUnit.MILLISECONDS);
    }


    public void shutdown() {
        try {
            scheduledExecutorService.awaitTermination(1, TimeUnit.DAYS);
        } catch (InterruptedException e) {
            e.printStackTrace();
        }
        try {
            executorService.awaitTermination(1, TimeUnit.DAYS);
        } catch (InterruptedException e) {
            e.printStackTrace();
        }
    }

}

File: core/src/main/kotlin/org/kantega/niagara/task/TaskExecutor.kt
package org.kantega.niagara.task


import arrow.core.Try
import java.time.Duration
import java.time.Instant
import java.util.concurrent.*

class TaskExecutor(
        internal val scheduledExecutorService: ScheduledExecutorService,
        internal val pool: ForkJoinPool,
        internal val defaultHandler: (Throwable) -> Task<Unit>) {

    fun <A> eval(task: Task<A>) {
        enqueue(
                TaskContext(this),
                task,
                { aTry -> aTry.fold({ Try { eval(defaultHandler(it)); it } }, { Try.just(it) }) })
    }

    fun <A> enqueue(tc: TaskContext, t: Task<A>, continuation: (Try<A>) -> Unit) {
        pool.submit {
            try {
                t.perform(tc, continuation)
            } catch (e: Throwable) {
                continuation(Try.raise(e))
            }
        }
    }

    fun <A> enqueueStage(tc: TaskContext, t: Task<A>, continuation: (Try<A>) -> Unit): CompletableFuture<Unit> {
        val cf = CompletableFuture<Unit>()
        pool.submit {
            try {
                t.perform(tc, continuation)
                cf.complete(Unit)
            } catch (e: Throwable) {
                continuation(Try.raise(e))
            }
        }
        return cf
    }

    fun <A> schedule(tc: TaskContext, t: Task<A>, continuation: (Try<A>) -> Unit, at: Instant) {
        val d = Duration.between(Instant.now(), at)
        if (d.isNegative)
            enqueue(tc, t, continuation)
        else
            scheduledExecutorService.schedule({ enqueue(tc, t, continuation) }, d.toMillis(), TimeUnit.MILLISECONDS)
    }


    fun shutdown() {
        try {
            scheduledExecutorService.awaitTermination(1, TimeUnit.DAYS)
        } catch (e: InterruptedException) {
            e.printStackTrace()
        }

        try {
            pool.awaitTermination(1, TimeUnit.DAYS)
        } catch (e: InterruptedException) {
            e.printStackTrace()
        }

    }

    companion object {

        var defaultExecutor =
                TaskExecutor(
                        Executors.newSingleThreadScheduledExecutor(),
                        ForkJoinPool.commonPool()
                ) { t -> Task { t.printStackTrace() } }
    }

}
